jack-core/src/com/rapleaf/jack/transaction/TransactorMetricsImpl.java

```diff
@@ -0,0 +1,99 @@
+package com.rapleaf.jack.transaction;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class TransactorMetricsImpl implements TransactorMetrics {
+  int longestQueriesListSize;
+
+
+  LinkedList<Long> longestTimes;
+
+  HashMap<StackTraceElement, Long> longestQueries;
+
+
+  //Contains a map with the *longestQueriesListSize* queries that took the most time to execute and their execution time
+
+
+  TransactorMetricsImpl(int longestQueriesSize) {
+    this.longestQueriesListSize = longestQueriesSize;
+    this.longestQueries = new HashMap<>();
+    this.longestTimes = new LinkedList<>();
+  }
+
+
+  synchronized void update(long executionTime, StackTraceElement queryStackTrace) {
+    if (longestQueries.containsKey(queryStackTrace)) {
+      if (longestQueries.get(queryStackTrace) < executionTime) {
+        longestTimes.removeFirstOccurrence(longestQueries.get(queryStackTrace));
```

Comment:
`longestTimes` is a linked list. So the `removeFirstOccurrence` needs O(N) time, in which N is the list size. This could be inefficient if the list size is large. I know that for now the list size is always 10, so that's fine. But can you try to think of a better data structure?